h/static/scripts/directive/annotation.js

```diff
@@ -0,0 +1,730 @@
+'use strict';
+
+var events = require('../events');
+
+/** Return truthy if the given annotation is valid, falsy otherwise.
+ *
+* A public annotation has to have some text and/or some tags to be valid,
+* public annotations with no text or tags (i.e. public highlights) are not
+* valid.
+*
+* Non-public annotations just need to have a target to be valid, non-public
+* highlights are valid.
+*
+* @param {object} annotation The annotation to be validated
+*
+*/
+function validate(annotation) {
+  var permissions;
+  var readPermissions;
+  var worldReadable = false;
+  var targets;
+
+  if (!angular.isObject(annotation)) {
+    return;
+  }
+
+  permissions = annotation.permissions || {};
+  readPermissions = permissions.read || [];
+  targets = annotation.target || [];
+
+  if (annotation.tags && annotation.tags.length) {
+    return annotation.tags.length;
+  }
+
+  if (annotation.text && annotation.text.length) {
+    return annotation.text.length;
+  }
+
+  if (readPermissions.indexOf('group:__world__') !== -1) {
+    worldReadable = true;
+  }
+
+  return (targets.length && !worldReadable);
+}
+
+/** Return a human-readable error message for the given server error.
+ *
+ * @param {object} reason The error object from the server. Should have
+ * `status` and, if `status` is not `0`, `statusText` and (optionally)
+ * `data.reason` properties.
+ *
+ * @returns {string}
+ */
+function errorMessage(reason) {
+  var message;
+  if (reason.status === 0) {
+    message = 'Service unreachable.';
+  } else {
+    message = reason.status + ' ' + reason.statusText;
+    if (reason.data.reason) {
+      message = message + ': ' + reason.data.reason;
+    }
+  }
+  return message;
+}
+
+/**
+  * @ngdoc type
```

Comment:
We don't ever compile it afaik, but we do have ngdoc (not just jsdoc) comments all over our code currently and it does have some [Angular-specific directives](https://github.com/angular/angular.js/wiki/Writing-AngularJS-Documentation#standard-supported-jsdoc-directives), particularly ones like `service`, `filter` and `directive`, that I think it might be worth keeping it for as long as we're using Angular.